### PR TITLE
Store start semester in URL

### LINKED ISSUE #43: Let users select their starting semester instead of their last completed semester
This would help in our future endeavor to validate the schedule based on when each module ran.
Also makes more sense to store your starting semester in the url than your last completed semester (#33). Since this always stays the same and does not automatically force you to update your saved bookmarks, etc. each semester.

### PR DESCRIPTION
**Pull request by jeremystucki**
Closes #43

### CODE REVIEW

File: src/components/Navigation.vue [hunk 1]
```diff
@@ -77,8 +82,7 @@ export default defineComponent({
...
             { title: 'Cybersecurity', path: 'RheKoI_OOP1_AutPy_CN1_DMI_An1I-FP_OOP2_DigCod_CySec_AutoSpr_An2I-WE1_AlgDat_Bsys1_AIFo_Dbs1_EnglScience-WE2_SecSoW_Bsys2_CN2_KommIng2-PmQm_PFSec_CldInf_NIoSec_SEP1_ExEv-CldOp_HackL_DSy_SEProj_SEP2_DigBusI-CyDef_AppArch_SAI21_WI2-IncResp_BAI21_PhAI' },
-            { title: 'Network and Cloud Infrastructure', path:
-            'RheKoI_OOP1_AutPy_CN1_DMI_An1I-FP_OOP2_DigCod_CySec_AutoSpr_An2I-WE1_AlgDat_Bsys1_AIFo_Dbs1_EnglScience-WE2_SModSim_DatAna_NetAut_CN2_KommIng2-PmQm_CldInf_NIoSec_SEP1_ExEv-CldOp_IBN_DSy_SEProj_SEP2_DigBusI-CyDef_AppArch_SAI21_WI2-CldSol_BAI21_PhAI' },
+            { title: 'Network and Cloud Infrastructure', path: 'RheKoI_OOP1_AutPy_CN1_DMI_An1I-FP_OOP2_DigCod_CySec_AutoSpr_An2I-WE1_AlgDat_Bsys1_AIFo_Dbs1_EnglScience-WE2_SModSim_DatAna_NetAut_CN2_KommIng2-PmQm_CldInf_NIoSec_SEP1_ExEv-CldOp_IBN_DSy_SEProj_SEP2_DigBusI-CyDef_AppArch_SAI21_WI2-CldSol_BAI21_PhAI' },
```
Comment (jeremystucki) [inline src/components/Navigation.vue:85]: I just removed the line break here

File: src/components/Navigation.vue [hunk 2]
```diff
@@ -89,8 +93,7 @@ export default defineComponent({
...
             { title: 'Cybersecurity', path: 'RheKoI_OOP1_AutPy_CN1_DMI_An1I_Dbs1_EnglScience-FP_OOP2_DigCod_CySec_AutoSpr_An2I_CN2-WE1_AlgDat_Bsys1_AIFo_CldInf_NIoSec_SEP1_ExEv-WE2_SecSoW_Bsys2_DSy_SEProj_SEP2_DigBusI-PmQm_PFSec_CyDef_AppArch_SAI21_WI2-CldOp_HackL_KommIng2_IncResp_BAI21_PhAI' },
-            { title: 'Network and Cloud Infrastructure', path:
-            'RheKoI_OOP1_AutPy_CN1_DMI_An1I_Dbs1_EnglScience-FP_OOP2_DigCod_CySec_AutoSpr_An2I_CN2-WE1_AlgDat_Bsys1_AIFo_CldInf_NIoSec_SEP1_ExEv-WE2_SModSim_DatAna_NetAut_DSy_SEProj_SEP2_DigBusI-PmQm_CyDef_AppArch_SAI21_WI2-CldOp_IBN_KommIng2_CldSol_BAI21_PhAI' },
+            { title: 'Network and Cloud Infrastructure', path: 'RheKoI_OOP1_AutPy_CN1_DMI_An1I_Dbs1_EnglScience-FP_OOP2_DigCod_CySec_AutoSpr_An2I_CN2-WE1_AlgDat_Bsys1_AIFo_CldInf_NIoSec_SEP1_ExEv-WE2_SModSim_DatAna_NetAut_DSy_SEProj_SEP2_DigBusI-PmQm_CyDef_AppArch_SAI21_WI2-CldOp_IBN_KommIng2_CldSol_BAI21_PhAI' },
```
Comment (jeremystucki) [inline src/components/Navigation.vue:96]: I just removed the line break here

File: src/components/Navigation.vue [hunk 3]
```diff
@@ -62,13 +62,18 @@
 
 <script lang="ts">
 import { defineComponent } from 'vue';
+import {SemesterInfo} from "../helpers/semester";
+
+const currentSemester = SemesterInfo.now();
+const defaultStartSemester = currentSemester.springSemester ? currentSemester.removeSemesters(1) : currentSemester;
```
Comment (StefanieJaeger) [inline src/components/Navigation.vue:68]: Moving the logic to find the most recent "Herbst Semester" to `SemesterInfo`, maybe behind a function called `getMostRecentFallSemester`, and calling it here would be an improvement.
Would make it easier to understand, what this code does and also keep all semester-related logic inside `SemesterInfo`.

File: src/helpers/semester.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
+const isSpringSemester = (date: Date) => {
+  const month = date.getMonth();
+  return month > 1 && month < 8;
+}
+
+export class SemesterInfo {
```
Comment (StefanieJaeger) [inline src/helpers/semester.ts:6]: Please call the file "semester-info.ts" or the like. Kinda confusing with "Semester.vue" also being a thing ^^'

File: src/views/Home.vue [hunk 5]
```diff
@@ -21,6 +21,8 @@
       :key="semester.number"
       v-model:modules="semester.modules"
       class="bg-gray-200 rounded p-2 group/semester w-64 min-w-64"
+      :title="startSemester === undefined
+        ? `${semester.number}` : startSemester.addSemesters(semester.number - 1).toString()"
```
Comment (StefanieJaeger) [inline src/views/Home.vue:25]:
```suggestion
      :title="startSemester ? startSemester.addSemesters(semester.number - 1).toString() : `${semester.number}`"
```

Comment (StefanieJaeger) [reply src/views/Home.vue:25]: And is "`${}`" and ".toString()" really needed here?

File: src/helpers/semester.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  readonly springSemester: boolean;
```
Comment (StefanieJaeger) [inline src/helpers/semester.ts:7]: Could we call this variable `isSprintSemester` or something similar? Just seeing the name, it is not clear that this is a boolean/what it is trying to tell me.

File: src/helpers/semester.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  toString() {
+    const prefix = this.springSemester ? "FS" : "HS";
+    const suffix = this.year.toString().substring(2);
+
+    return prefix + suffix;
+  }
+
+  addSemesters(count: number) {
```
Comment (StefanieJaeger) [inline src/helpers/semester.ts:48]: I really dislike the naming `addSemesters` and `removeSemester`...
Could we maybe get a static `increaseSemesters(count: number, baseSemester: SemesterInfo)` and static `decreaseSemesters(count: number, baseSemester: SemesterInfo)`? Small change, but feels much cleaner.

File: src/helpers/semester.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    if (other === undefined) return false;
+
+    if (other instanceof SemesterInfo) {
+      return this.springSemester === other.springSemester && this.year === other.year;
+    }
+
+    return false;
+  }
+
+  difference(other: SemesterInfo) {
```
Comment (StefanieJaeger) [inline src/helpers/semester.ts:83]: Watch out, this could return negative. Is that wanted/considered for all usages?

**Review by StefanieJaeger — CHANGES_REQUESTED** [review]

File: src/views/Home.vue [hunk 9]
```diff
@@ -234,7 +264,18 @@ export default defineComponent({
           .replace('NISec', 'NIoSec')
           .replace('PFSec', 'PlFSec');
 
-        const planData = newPath
+        const [ hash, query ] = newPath.split('?');
+
+        if (query != undefined) {
```
Comment (StefanieJaeger) [inline src/views/Home.vue:269]: If url query does not contain "startSemester", we just skip this code. But shouldn't we make sure to reset the selected semester? Or write the selected semester to the url.

**Review by StefanieJaeger — CHANGES_REQUESTED** [review]

File: src/components/Navigation.vue (hunk 3 above)
Comment (jeremystucki) [reply src/components/Navigation.vue:68]: I don't have time for this. Feel free to do so.

File: src/views/Home.vue (hunk 5 above)
Comment (jeremystucki) [reply src/views/Home.vue:25]: Yes

File: src/views/Home.vue (hunk 9 above)
Comment (StefanieJaeger) [reply src/views/Home.vue:269]: This is only a problem, when manually editing the url (eg removing startSemester) and then pressing enter. Refreshing or loading with an url not containing that query param does not pose a problem.

File: src/components/Navigation.vue (hunk 3 above)
Comment (jeremystucki) [reply src/components/Navigation.vue]: I changed it now, found the motivation somewhere.

File: src/helpers/semester.ts (hunk 7 above)
Comment (jeremystucki) [reply src/helpers/semester.ts]: I called it `plus` and `minus` now.

**Review by StefanieJaeger — APPROVED** [review]

**Pull request merged**